[Remarks] Introduce tool to count remarks.

zjaffal opened the pull request:
This tool is a generic remark counter reporting count based on specified properties. The counter can be used to count remarks individually and filter them based on name, type and pass or count using remark arguments.

Commit:
Remark Util introduce remark count

This tool is a generic remark counter reporting count based on specified
properties. The counter can be used to count remarks individually and
filter them based on name, type and pass or count using remark
arguments.

fhahn commented on llvm/tools/llvm-remarkutil/RemarkCounter.h:
```diff
@@ -0,0 +1,198 @@
+/// by reporting count for functions, source or total count for the provided
+/// remark file.
+struct RemarkCounter : Counter {
+  std::map<std::string, int> CountedByRemarksMap;
```
Does this map rely on being ordered by the string order?

zjaffal replied:
it makes sense for the tool to report the remark count by the order it parsed the files in

fhahn commented on llvm/tools/llvm-remarkutil/RemarkCounter.h:
```diff
@@ -0,0 +1,198 @@
+struct RemarkCounter : Counter {
+  std::map<std::string, int> CountedByRemarksMap;
+  RemarkCounter(enum GroupBy GroupBy) : Counter(GroupBy) {}
+  void collect(const Remark &) override;
```
Document?

fhahn replied:
still looks unresolved?

fhahn reviewed (commented):
Overall this looks great, thanks for the detailed documentation!

A bunch of suggestions inline

zjaffal commented on llvm/tools/llvm-remarkutil/RemarkCounter.cpp:
```diff
@@ -0,0 +1,339 @@
+  });
+  if (RemarkArg == Remark.Args.end())
+    return 0;
+  return *RemarkArg->getValAsInt();
```
If no remark is found then we return 0.

fhahn replied:
Got it, thanks!

zjaffal commented on llvm/tools/llvm-remarkutil/RemarkCounter.cpp:
```diff
@@ -0,0 +1,339 @@
+
+Error Filters::regexArgumentsValid() {
+  if (RemarkNameFilter.has_value() && RemarkNameFilter->IsRegex)
+    if (auto E = checkRegex(std::get<Regex>(RemarkNameFilter->FilterRE)))
```
Regex returns a string error so `checkRegex` here converts that string error into an error object. We need to do the check before we use the regex objects.

fhahn replied:
Got it, thanks!

Commit:
[Remark] Remove unrelated change for Remark header.

Commit:
Update remark.rst file

Commit:
Address nit comments from PR

1. Change the description for command line options to be more consistent
2. Fix doumentation issues
3. Change argumetns to use ArrayRef
4. Use unsigned instead of signed when reporting count.
5. Simplify some of the code by using `any_of`

francisvm commented on llvm/docs/CommandGuide/llvm-remarkutil.rst:
```diff
@@ -110,6 +110,80 @@ if `--use-debug-loc` is passed then the CSV will include the source path, line n
+Summary
+^^^^^^^
+
+:program:`llvm-remarkutil count` counts `remakrs <https://llvm.org/docs/Remarks.html>` based on specified properties. 
```
```suggestion
:program:`llvm-remarkutil count` counts `remarks <https://llvm.org/docs/Remarks.html>` based on specified properties. 
```

fhahn replied:
Looks like the typo is still there in the latest version?

francisvm commented on llvm/docs/CommandGuide/llvm-remarkutil.rst:
```diff
@@ -110,6 +110,80 @@ if `--use-debug-loc` is passed then the CSV will include the source path, line n
+  Filter count by remark name.
+
+.. option:: --rremark-name[=<string>]
+  Filter count by remark name using regular expressions.
```
Why have both name and regex for these options? Isn't the regex enough?

zjaffal replied:
It is just following the same idea that other tools have. For example `llvm-extract` has `--func` and `--rfunc`, I think if you pass something like `remark-name` to the llvm regex parser it will also match with `remark-name1`, `remark-name2` so it is better to seperate regex option

fhahn replied:
I think it should be fine to have them separately initially. Discussed this with Zain offline and having a separate `remark-name` makes it easier to match names exactly; with one option that takes a regex only the user would have to write a more complicated regex

zjaffal commented on llvm/docs/CommandGuide/llvm-remarkutil.rst:
```diff
@@ -110,6 +110,80 @@ if `--use-debug-loc` is passed then the CSV will include the source path, line n
+  group count of remarks by property.
+  * ``source``: Count will be collected per source path. Requires remarks to have debug loc info
+  * ``function``: Count is collected per function.
+  * ``function-with-loc``: Count is collected per function per source. Requires remarks to have debug loc info 
```
Yes I will update it accordingly

fhahn replied:
Still needs updating I think

fhahn commented on llvm/tools/llvm-remarkutil/RemarkCounter.cpp:
```diff
@@ -0,0 +1,336 @@
+// Generic tool to count remarks based on properties
+//
+//===----------------------------------------------------------------------===//
+#include "RemarkCounter.h"
```
nit: newline above

Commit:
Rename `KeyCounter` to `ArgumentCounter`

The class and the subsiquent variables are used to collect count
information from remark arguments. It is easier for any future
contributers and users of the count tool to get that when the naming
align with the specified function

Commit:
Update documentation for functions and nit comments from PR

fhahn reviewed (approved):
LGTM, thanks!

fhahn reviewed (changes requested):
Just noticed there are some build failures: https://buildkite.com/llvm-project/github-pull-requests/builds/6840#018b196d-9977-4b10-abf6-002c37b825ae

`std::variant` is only available in c++17, but LLVM defaults to c++14, so it cannot be used.

dlumma commented:
Direct link to failure: https://buildkite.com/llvm-project/github-pull-requests/builds/6840#018b196d-9977-4b10-abf6-002c37b825ae/6-2264

Commit:
Remove std::variant from FilterMatcher class

std::variant is available in c++17 but LLVM defaults to c++14.

fhahn reviewed (approved):
LGTM, thanks!

jplehr commented:
This appears to have broken the AMDGPU OpenMP buildbot https://lab.llvm.org/buildbot/#/builders/193/builds/39996 and the Clang HIP buildbot https://lab.llvm.org/buildbot/#/builders/165/builds/44145

AaronBallman commented:
There are still broken bots from this change: https://lab.llvm.org/buildbot/#/builders/124/builds/8595 @zjaffal, please revert the changes until you have a fix ready.

MaskRay commented:
> There are still broken bots from this change: [lab.llvm.org/buildbot/#/builders/124/builds/8595](https://lab.llvm.org/buildbot/#/builders/124/builds/8595) @zjaffal, please revert the changes until you have a fix ready.

The GCC 7.5 build issue was due to `return Filter;`. I fixed it and the reserved identifier `_GroupBy` (introduced by a follow-up c136e722aa4b03209da48b641c6f413202cb0ff9 to this patch) in 0603737ac0f73ff33326d0274b3152cc5a81e1ec.

It would be cleaner the revert the patch, but there are so many follow-up commits (and would cause conflicts), so I added another follow-up instead of reverting the whole patch.

AaronBallman commented:
> > There are still broken bots from this change: [lab.llvm.org/buildbot/#/builders/124/builds/8595](https://lab.llvm.org/buildbot/#/builders/124/builds/8595) @zjaffal, please revert the changes until you have a fix ready.
> 
> The GCC 7.5 build issue was due to `return Filter;`. I fixed it and the reserved identifier `_GroupBy` (introduced by a follow-up [c136e72](https://github.com/llvm/llvm-project/commit/c136e722aa4b03209da48b641c6f413202cb0ff9) to this patch) in [0603737](https://github.com/llvm/llvm-project/commit/0603737ac0f73ff33326d0274b3152cc5a81e1ec).
> 
> It would be cleaner the revert the patch, but there are so many follow-up commits (and would cause conflicts), so I added another follow-up instead of reverting the whole patch.

Thank you! A fix forward is also wonderful. :-)
